articles/141_static_remapping.md

```diff
@@ -0,0 +1,435 @@
+layout: default
+title: Remapping Names
+permalink: articles/static_remapping.html
+abstract:
+  Topics, parameters, and services are identified by [Names](http://wiki.ros.org/Names). Names are hard coded in ROS nodes, but they can be changed at runtime through remapping. Without remapping every instance of a node would require changes in code. This article describes the requirements, rationale, and mechanisms for remapping names in ROS 2.
+author: '[Shane Loretz](https://github.com/sloretz)'
+published: true
+---
+
+- This will become a table of contents (this text will be scraped).
+{:toc}
+
+# {{ page.title }}
+
+<div class="abstract" markdown="1">
+{{ page.abstract }}
+</div>
+
+Original Author: {{ page.author }}
+
+## Why remap names?
+Remapping names allows reusing the same node executable in different parts of the system.
+A robot that has multiple sensors of the same type could launch multiple instances of the same node with outputs remapped to different topics.
+
+```
+            +-------------+
+ (Lidar 1)--> node inst 1 +-->/head_scan
+            +-------------+
+
+            +-------------+
+ (Lidar 2)--> node inst 2 +-->/base_scan
+            +-------------+
+```
+
+
+## Structure of a Name
+Names are hierarchical strings delineated by `/`.
+If they begin with `/` they are said to be **global** or **Fully Qualified Names** (FQN).
+Otherwise they are said to be **relative**.
+Relative names are resolved to FQN prior to being used.
+
+**Example names:**
+
+- `/foo`
+- `/foo/bar`
+- `foo/bar/baz`
+- `foo/bar`
+- `bar`
+
+Each part of a name between a slash is a **token**.
+The examples above use the tokens `foo`, `bar` and `baz`.
+FQN are conceptually divided into two pieces: **namespace** and **basename**.
+The basename is the last token in a name.
+The namespace is all tokens and slashes prior to the basename.
+
+**Example namespaces:**
+
+- `/`
+- `/foo/`
+- `/foo/bar/baz/`
+
+**Example basenames:**
+
+- `bar`
+- `foo`
+
+Nodes are said to be in a **namespace**.
+This namespace becomes a prefix to all relative names used by the node.
+Relative names are expanded to FQN by prepending the node's namespace to them.
+
+
+**Example:**
+
+- Node is in namespace `/fiz/buz`
+- Node uses relative name `foo/bar`
+- The FQN is `/fiz/buz/foo/bar`
+
+
+## Structure of a Remapping Rule
+**Remapping rules** are the instructions describing how a node should change the names it uses.
+Remapping rules have two parts.
+The first part is used to determine if the rule applies to a name.
+The second part is the replacement for a matched name.
+The act of replacing one name with another is **remapping**.
+
+
+## ROS 2 Remapping Use cases
+These use cases are being considered for remapping in ROS 2:
+
+- Remap One Node in a Process
+- Change a Namespace
+- Change a Basename
+- Change a Token
+- Pre-FQN Remapping
+- Exact FQN Replacement
+- Exact Relative Name Replacement
+- Remap via Command Line
+- Change the Default Namespace
+- Change the Node Name
```

Comment:
I added a section about the order of remapping rules anyways. It was an opportunity to add that rules stack. It's pointless for static remapping, but important for dynamic remapping the visible name.
8cff468696f1881ada90c30a75a88e26e7eee0ee